proto/engine/v1/json_marshal_unmarshal.go

```diff
@@ -238,6 +277,83 @@ func (e *ExecutionPayload) UnmarshalJSON(enc []byte) error {
 	return nil
 }
 
+// TODO: Refactor this to not duplicate the entirety of ExecutionPayload's UnmarshalJSON
+func (e *ExecutionPayload4844) UnmarshalJSON(enc []byte) error {
+	dec := executionPayloadJSON{}
+	if err := json.Unmarshal(enc, &dec); err != nil {
+		return err
+	}
+
+	if dec.ParentHash == nil {
+		return errors.New("missing required field 'parentHash' for ExecutionPayload")
+	}
+	if dec.FeeRecipient == nil {
+		return errors.New("missing required field 'feeRecipient' for ExecutionPayload")
+	}
+	if dec.StateRoot == nil {
+		return errors.New("missing required field 'stateRoot' for ExecutionPayload")
+	}
+	if dec.ReceiptsRoot == nil {
+		return errors.New("missing required field 'receiptsRoot' for ExecutableDataV1")
+	}
+
+	if dec.LogsBloom == nil {
+		return errors.New("missing required field 'logsBloom' for ExecutionPayload")
+	}
+	if dec.PrevRandao == nil {
+		return errors.New("missing required field 'prevRandao' for ExecutionPayload")
+	}
+	if dec.ExtraData == nil {
+		return errors.New("missing required field 'extraData' for ExecutionPayload")
+	}
+	if dec.BlockHash == nil {
+		return errors.New("missing required field 'blockHash' for ExecutionPayload")
+	}
+	if dec.Transactions == nil {
+		return errors.New("missing required field 'transactions' for ExecutionPayload")
+	}
+	if dec.BlockNumber == nil {
+		return errors.New("missing required field 'blockNumber' for ExecutionPayload")
+	}
+	if dec.Timestamp == nil {
+		return errors.New("missing required field 'timestamp' for ExecutionPayload")
+	}
+	if dec.GasUsed == nil {
+		return errors.New("missing required field 'gasUsed' for ExecutionPayload")
+	}
+	if dec.GasLimit == nil {
+		return errors.New("missing required field 'gasLimit' for ExecutionPayload")
+	}
+	if dec.ExcessBlobs == nil {
```

Comment:
I don't know how the EL knows to upgrade to EIP4844 and start supplying ExcessBlobs. Ideally the CL would inform the EL to provide 4844 ExecutionData, and then this logic would be correct.

If the EL can't be trusted to always upgrade to include the 4844 ExecutionPayload when the CL starts requesting it... that's a problem. We might default to zero, but then the hash will not match?

What do you think should be done?

Reply:
For now let's assume the CL can handle 4844 Payloads before the EL upgrade.

Technically, the CL doesn't request a 4844-specific payload, it's happy with whatever type of payload returned by the EL. If the EL builds and returns a 4844 block then the CL should simply add the unadulterated payload to the chain via `NewPayload`. Same reasoning applies for pre-4844 blocks. 
But this can only work if the CL is able to handle 4844 blocks a priori (via a CL hardfork).

So I think the nil check for ExcessBlobs should be removed here. The nil checks for the pre-4844 fields are essential though because those are the base attributes of a payload. But as we upgrade the ExecutionPayload header in future forks, I think it's easiest to simply treat header extension fields as _optional_.

We can always change this behavior later if client devs feel otherwise. But for now let's treat execution payloads as opaque as possible in the CL and not do nil checks on new header fields.